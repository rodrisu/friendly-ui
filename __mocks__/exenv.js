// Mocks the exenv library and makes its values mutable

module.exports = {
  canUseDOM: true,
  canUseEventListeners: true,
}
