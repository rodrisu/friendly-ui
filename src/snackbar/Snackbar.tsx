import React, { PureComponent } from 'react'
import { createPortal } from 'react-dom'
import TransitionGroup from 'react-transition-group/TransitionGroup'
import cc from 'classcat'
import { canUseDOM } from 'exenv'

import { color } from '../_utils/branding'
import { Button, ButtonStatus } from '../button'
import { CrossIcon } from '../icon/crossIcon'
import { Text, TextDisplayType } from '../text'
import { AnimationType, Transitions as CustomTransition } from '../transitions'
import { StyledSnackbar } from './Snackbar.style'

export type SnackbarProps = Readonly<{
  close: () => void
  isOpen: boolean
  className?: string
  extraClassName?: string
  children: React.ReactNode
}>

export class Snackbar extends PureComponent<SnackbarProps> {
  private portalNode: HTMLElement
  static defaultProps: Partial<SnackbarProps> = {
    isOpen: false,
  }

  constructor(props: SnackbarProps) {
    super(props)
    if (canUseDOM) {
      this.portalNode = document.createElement('div')
      document.body.appendChild(this.portalNode)
    }
  }

  componentWillUnmount() {
    if (this.portalNode) {
      document.body.removeChild(this.portalNode)
      this.portalNode = null
    }
  }

  render() {
    const modalElement = (
      <TransitionGroup
        component={StyledSnackbar}
        className={cc(['transition-wrapper kirk-snackbar-container', this.props.className])}
      >
        {this.props.isOpen && (
          <CustomTransition animationName={AnimationType.SLIDE_UP}>
            <div className={cc(['kirk-snackbar', this.props.extraClassName])}>
              <Text role="alert" display={TextDisplayType.TITLE} className="kirk-snackbar-content">
                {this.props.children}
              </Text>
              <Button
                status={ButtonStatus.UNSTYLED}
                className="kirk-snackbar-cross"
                onClick={this.props.close}
              >
                <CrossIcon iconColor={color.white} size="18px" />
              </Button>
            </div>
          </CustomTransition>
        )}
      </TransitionGroup>
    )

    if (!this.portalNode) {
      return modalElement
    }

    return createPortal(modalElement, this.portalNode)
  }
}
