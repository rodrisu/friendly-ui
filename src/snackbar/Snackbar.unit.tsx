import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import exenv from 'exenv'

import { fireEvent, render, screen } from '@testing-library/react'

import { Snackbar as StyledSnackbar } from './index'
import { Snackbar } from './Snackbar'

const close = jest.fn()

describe('Snackbar', () => {
  it('should not have changed', () => {
    exenv.canUseDOM = false
    const snackbarServerSide = renderer
      .create(
        <StyledSnackbar isOpen close={close}>
          Oups
        </StyledSnackbar>,
      )
      .toJSON()
    exenv.canUseDOM = true
    expect(snackbarServerSide).toMatchSnapshot()
  })

  it('should show the snackbar when isOpen is true', () => {
    render(
      <Snackbar isOpen close={close}>
        Oups
      </Snackbar>,
    )

    const snackbar = screen.getByRole('alert')
    expect(snackbar).toBeInTheDocument()
    expect(snackbar).toHaveTextContent('Oups')
  })

  it('should npt show the snackbar when isOpen is false', () => {
    render(
      <Snackbar isOpen={false} close={close}>
        Oups
      </Snackbar>,
    )

    const snackbar = screen.queryByRole('alert')
    expect(snackbar).not.toBeInTheDocument()
  })

  it('should call close method when click on cross button', () => {
    render(<Snackbar isOpen close={close} />)

    fireEvent.click(screen.getByRole('button'))

    expect(close).toHaveBeenCalled()
  })

  it('should render the same layout on client and server side', () => {
    exenv.canUseDOM = false
    const serverSide = mount(<Snackbar isOpen close={close} />)
    exenv.canUseDOM = true
    const clientSide = mount(<Snackbar isOpen close={close} />)

    expect(serverSide.html()).toEqual(clientSide.html())
  })
})
