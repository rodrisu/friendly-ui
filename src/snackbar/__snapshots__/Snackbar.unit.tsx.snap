// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snackbar should not have changed 1`] = `
.c4 {
  fill: #FFF;
}

.c4.kirk-icon-wrapper {
  display: inline-block;
  position: relative;
}

.c4.kirk-icon-shadowed {
  border-radius: 50%;
  box-shadow: 0 0 2px rgba(0,0,0,.3);
}

.c4 .kirk-icon-badge {
  position: absolute;
  bottom: 12px;
  left: 12px;
}

.c3 {
  position: relative;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding: 0 24px;
  border: 1px solid transparent;
  border-radius: 48px;
  font-size: 16px;
  line-height: 1;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
  font-family: inherit;
  height: 48px;
  min-width: 48px;
  max-width: 100%;
  overflow: hidden;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-transition: max-width 500ms ease-in, background-color 200ms ease-in;
  transition: max-width 500ms ease-in, background-color 200ms ease-in;
  white-space: nowrap;
}

.c3 svg,
.c3 img {
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.c3 svg {
  margin-right: 8px;
}

.c3.kirk-button-bubble svg {
  margin: 0;
}

.c3.kirk-itemChoice {
  border-radius: 0;
  overflow: visible;
  width: 100%;
}

.c3:hover,
.c3:active {
  outline: 0;
}

.c3.kirk-button-unstyled {
  background-color: transparent;
  border-radius: inherit;
  color: #00AFF5;
  padding: inherit;
  display: inherit;
  line-height: inherit;
  font-size: inherit;
  text-align: left;
  min-width: inherit;
  min-height: inherit;
  height: auto;
  white-space: normal;
}

.c3[disabled] {
  opacity: 0.5;
  cursor: default;
}

.c3.kirk-button-unstyled[disabled] {
  opacity: 1;
}

.c3.kirk-button-primary {
  background-color: #00AFF5;
  color: #FFF;
}

.c3.kirk-button-shadowed {
  border: 2px solid #FFF;
  box-shadow: 0 2px 4px rgba(0,0,0,.3);
  box-sizing: border-box;
}

.c3.kirk-button-primary:hover,
.c3.kirk-button-primary:focus,
.c3.kirk-button-primary:active {
  background-color: #008FC1;
}

.c3.kirk-button-secondary {
  background-color: #FFF;
  border-color: #DDD;
  color: #00AFF5;
}

.c3.kirk-button-secondary.kirk-button-shadowed {
  border: none;
}

.c3.kirk-button-secondary:hover,
.c3.kirk-button-secondary:focus,
.c3.kirk-button-secondary:active {
  background-color: #DDD;
}

.c3.kirk-button-tertiary {
  background-color: #FFF;
  border-color: transparent;
  color: #00AFF5;
}

.c3.kirk-button-tertiary:hover,
.c3.kirk-button-tertiary:focus,
.c3.kirk-button-tertiary:active {
  background-color: #DDD;
}

.c3.kirk-button-loading {
  background-color: transparent;
  max-width: initial;
  border: 0;
}

.c3.kirk-button-loading:hover,
.c3.kirk-button-loading:focus,
.c3.kirk-button-loading:active {
  background-color: transparent;
}

.c3.kirk-button-loading[disabled],
.c3.kirk-button-checked[disabled] {
  opacity: 1;
}

.c3 svg.kirk-button-loader {
  position: absolute;
  top: 0;
  left: 50%;
  margin-left: -24px;
}

.c3.kirk-button-checked {
  border: 0;
}

.c3.kirk-button-warning {
  background-color: #F53F5B;
  color: #FFF;
}

.c3.kirk-button-bubble {
  display: inline-block;
  text-align: center;
  padding: 0;
  line-height: 0;
  width: 48px;
  height: 48px;
}

.c3.kirk-button-bubble.kirk-button-shadowed {
  width: calc(48px - 2px * 2);
  height: calc(48px - 2px * 2);
}

.c3.kirk-button-facebook {
  background-color: #1877F2;
  color: #FFF;
}

.c3.kirk-button-vkontakte {
  background-color: #2787F5;
  color: #FFF;
}

.c2 {
  margin: 0;
  font-weight: 400;
}

.c2.kirk-text-button {
  color: #054752;
  font-size: 16px;
  line-height: 20px;
}

.c2.kirk-text-body,
.c2.kirk-text-bodyStrong {
  color: #708C91;
  font-size: 16px;
  line-height: 20px;
}

.c2.kirk-text-caption {
  color: #708C91;
  font-size: 13px;
  line-height: 16px;
}

.c2.kirk-text-title,
.c2.kirk-text-titleStrong {
  color: #054752;
  font-size: 18px;
  line-height: 20px;
}

.c2.kirk-text-display2 {
  color: #054752;
  font-size: 82px;
  line-height: 82px;
  font-weight: 500;
}

.c2.kirk-text-display1 {
  color: #054752;
  font-size: 30px;
  line-height: 1.06;
  font-weight: 500;
}

.c2.kirk-text-subheader,
.c2.kirk-text-subheaderStrong {
  color: #054752;
  font-size: 22px;
  line-height: 24px;
}

.c2.kirk-text-title,
.c2.kirk-text-titleStrong {
  color: #054752;
  font-size: 18px;
  line-height: 20px;
}

.c2.kirk-text-bodyStrong,
.c2.kirk-text-subheaderStrong,
.c2.kirk-text-titleStrong {
  font-weight: 500;
}

.c1.fade {
  -webkit-transition: ease-in-out 200ms;
  transition: ease-in-out 200ms;
  -webkit-transition-property: opacity;
  transition-property: opacity;
}

.c1.fade-entering {
  opacity: 0;
}

.c1.fade-entered {
  opacity: 1;
}

.c1.fade-exiting {
  opacity: 0;
}

.c1.slide-up {
  -webkit-transition: ease-in-out 200ms;
  transition: ease-in-out 200ms;
  -webkit-transition-property: opacity,-webkit-transform;
  -webkit-transition-property: opacity,transform;
  transition-property: opacity,transform;
}

.c1.slide-up-entering {
  opacity: 0;
  -webkit-transform: translateY(50%);
  -ms-transform: translateY(50%);
  transform: translateY(50%);
}

.c1.slide-up-entered {
  opacity: 1;
}

.c1.slide-up-exiting {
  opacity: 0;
  -webkit-transform: translateY(50%);
  -ms-transform: translateY(50%);
  transform: translateY(50%);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: fixed;
  bottom: 0;
  left: 0;
  right: 0;
  z-index: 1000;
}

.c0 .kirk-snackbar {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-self: center;
  -ms-flex-item-align: center;
  align-self: center;
  width: 100%;
  max-width: 614px;
  overflow: hidden;
  background-color: #F53F5B;
  padding: 16px 24px;
}

.c0 .kirk-snackbar .kirk-snackbar-content {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  margin: 0;
  color: #FFF;
}

.c0 .kirk-snackbar .kirk-snackbar-cross {
  padding: 0 0 0 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
}

@media (hover:none),(hover:on-demand) {
  .c3.kirk-button-secondary:hover {
    background-color: #FFF;
  }
}

<div
  className="c0 transition-wrapper kirk-snackbar-container"
>
  <div
    className="c1 kirk-snackbar slide-up slide-up-entered"
  >
    <span
      className="kirk-text kirk-text-title c2 kirk-snackbar-content"
      role="alert"
    >
      Oups
    </span>
    <button
      className="kirk-button kirk-button-unstyled c3 kirk-snackbar-cross"
      disabled={false}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchStart={[Function]}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="kirk-icon c4"
        height="18px"
        viewBox="0 0 24 24"
        width="18px"
        xmlns="http://www.w3.org/2000/svg"
      >
        
        <path
          d="M19 5L5 19M19 19L5 5"
          fill="none"
          stroke="#FFF"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeMiterlimit="10"
          strokeWidth="2"
        />
      </svg>
    </button>
  </div>
</div>
`;
